File: Tree & Binary Tree/Binary Tree Preorder Traversal/solution_iterative.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def preorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        stack = []
        current = root
        result = []
        while current or stack:
            if current:
                result.append(current.val)
                if current.right:
                    stack.append(current.right)
                current = current.left
            else:
                current = stack.pop()
        return result

File: Tree & Binary Tree/Binary Tree Preorder Traversal/solution_iterative.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<int> preorderTraversal(TreeNode* root) {
        if(root == nullptr) return {};
        stack<TreeNode*> st;
        st.push(root);
        vector<int> result;
        while(st.size() > 0) {
            TreeNode* node = st.top();
            st.pop();
            result.push_back(node->val);
            if(node->right != nullptr) {
                st.push(node->right);
            }
            if(node->left != nullptr) {
                st.push(node->left);
            }
        }
        return result;
    }
};
